application/common/models/User.php

```diff
@@ -40,6 +48,195 @@ public function rules()
+            );
+        }
+
+        return $user;
+    }
+
+
+    /**
+     * Update local user record if given properties are different than currently stored
+     * @param string $firstName
+     * @param string $lastName
+     * @param string $username
+     * @param string $email
+     * @return bool True if profile was updated, false if not
+     * @throws \Exception
+     */
+    public function updateProfileIfNeeded($firstName, $lastName, $username, $email)
+    {
+        $dirty = false;
+        $properties = [
+            'first_name' => $firstName,
+            'last_name' => $lastName,
+            'idp_username' => $username,
+            'email' => $email,
+        ];
+
+        foreach ($properties as $property => $value) {
+            if ($this->$property != $value) {
+                $dirty = true;
+                $this->$property = $value;
+            }
+        }
+
+        if ($dirty) {
+            if ($this->save()) {
+                return true;
+            } else {
+                /**
+                 * @todo add logging with model validation errors
+                 */
+                throw new \Exception('Unable to update profile', 1456760819);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Return array of arrays of masked out methods
+     * @return array
+     */
+    public function getMaskedMethods()
+    {
+        $methods = [];
+        foreach ($this->methods as $method) {
+            $methods[] = [
+                'uid' => $method->uid,
+                'type' => $method->type,
+                'value' => $method->getMaskedValue(),
+            ];
+        }
+        return $methods;
+    }
+
+    /**
+     * @return PersonnelUser
+     * @throws \Exception
+     */
+    public function getPersonnelUser()
+    {
+        if ( ! empty($this->personnelUser)) {
+            return $this->personnelUser;
+        }
+
+        /*
+         * If session is available, cache data in session
+         */
+        try {
+            $sessionAvailable = (\Yii::$app->user->identity == $this);
+        } catch (\Exception $e) {
+            $sessionAvailable = false;
+        }
+
+        if ($sessionAvailable && is_array(\Yii::$app->session->get('personnelUser'))) {
+            return \Yii::$app->session->get('personnelUser');
+        }
+
+        /*
+         * Fetch data from Personnel system and cache it
+         */
+        $this->fetchPersonnelUser();
+        \Yii::$app->session->set('personnelUser', $this->personnelUser);
+
+        return $this->personnelUser;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasSupervisor()
+    {
+        $personnelUser = $this->getPersonnelUser();
+        return $personnelUser->supervisorEmail !== null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasSpouse()
+    {
+        $personnelUser = $this->getPersonnelUser();
+        return $personnelUser->spouseEmail !== null;
```

Comment:
Not that it's much code, but we could slightly reduce duplication here by having this `hasSpouse()` function use the `getSpouseEmail()` function internally, since they do almost exactly the same thing.  Same with the supervisor functions here.  Just a thought.